Added support for renaming duplicate headers

fortydegrees opened the pull request:
Added code to automatically rename headers to solve the issue in #129 

I tried to add a renameDuplicateHeaders config option but left it out as I am doing the processing in the Parse.parse function where I don't have the ability to read the config. I'm not sure it fits the styling of the codebase to modify the input outside of that function so have left it out.

fortydegrees commented:
Thanks for the quick feedback & sorry for the misunderstanding re: test cases.

I've added a test case, removed the useless comments, and added a conditional based on whether the header is present. 

You were right - the renaming code could be refactored. I originally just took the code from that issue thread which worked, but rewrote it in a simpler way which produces the same result.

pokoli commented on papaparse.js:
```diff
@@ -1457,6 +1457,34 @@ License: MIT
+				let header_map = []; 
+				let header_count = {};
+
+				for (var i in headers){
```
I guess it will be easier to read if you write: 

var header = headers[i]; 

and then use `header` instead of `headers[i]` in all of the loop code.

pokoli commented on papaparse.js:
```diff
@@ -1457,6 +1457,34 @@ License: MIT
+			{
+				let firstLine = input.split(newline)[0]
+				let separator = '_';
+				let headers = firstLine.split(delim) 
```
Could you add `headers`variable after `firstLine`? As they are linked it make sense to define one after the other.

pokoli commented on papaparse.js:
```diff
@@ -1457,6 +1457,34 @@ License: MIT
+						else{
+							header_count[headers[i]] +=1
+						}
+						header_map.push(headers[i]+separator+header_count[headers[i]])
```
I will prefer to have a single line that pushes to `header_map`.
So you just need to have a variable name `header_name` that is initialized with the same value as `header`

Then if `header` is already in `header_map+  you just save the count and update the variable name.

pokoli commented on tests/test-cases.js:
```diff
@@ -585,6 +585,15 @@ var CORE_PARSER_TESTS = [
 		}
+	},
+	{
+		description: "Duplicate header names",
```
Could you add a test with also transformHeader function? As we are "manipulating" the headers I think we need to ensure that both functionalities work together.

pokoli reviewed (changes requested):
Thanks for your work on this. Everything looks nice, I just added some comments to improve code redability and ensure everything works as expected. Could you please review them?

fortydegrees commented:
Have addressed all comments re: code-cleanup.

Good call on the transformHeader() test. I added it and It originally failed. I added a check for the function when initialising the header in my code. Hopefully this is satisfactory.

I also added an extra test in the case of an existing header matching an intended substitution. An unlikely situation but thought it'd be wise to cover it.

pokoli commented on tests/test-cases.js:
```diff
@@ -585,7 +585,34 @@ var CORE_PARSER_TESTS = [
+		}
+	},
+	{
+		description: "Duplicate header names where an original header matches intended substitution",
```
I do not have a clear way that this is the best way to do. Probably we should respect the original names, but this makes code more complex.

As this is a corner case I think the best is to remove such test, and let the user decide what to do by using the transformFunction confiuguration.

Otherwise if we keep the test we are agreeing that this is the expected behaviour and I do not think is the right way.

fortydegrees replied:
I agree. Have removed the test.

pokoli commented on papaparse.js:
```diff
@@ -1457,6 +1457,37 @@ License: MIT
+						header = config.transformHeader(header, i);
+					var header_name = header
+					if (header_map.includes(header_name)){
+						if (!(header in header_count)){
```
I think its simpler to write as:

count = header_count[header_name] || '1';
header_name = header+separator+count;
header_count[header_name] = count +1;

Removeing the if else lines.

fortydegrees replied:
I struggled to make it work without an if statement. The problem being that it would still add the suffix at the end. The best I got was:

`let count = header_count[header] || 0
if (count > 0) header_name = header+separator+count;
header_count[header] = count+1`

Have committed this for you consideration if you prefer.

pokoli reviewed (changes requested):
Everythinkg looks nice now. I leaved some minor issues that should be addressed.

Could you fix them? I think we will be able to merge it once they are addressed. 

Thanks for your time on this.

pokoli commented on papaparse.js:
```diff
@@ -1457,6 +1457,32 @@ License: MIT
+			// Rename headers if there are duplicates
+			if (config.header)
+			{
+				let firstLine = input.split(newline)[0]
```
please add ; as ending char on all lines to follow the same style.

pokoli commented on papaparse.js:
```diff
@@ -1457,6 +1457,32 @@ License: MIT
+
+					header_map.push(header_name)
+				}
+				var edited_input = input.split(newline)
```
We should avoid spliting the input when the computed header is the same as the previous one. 
This will avoid some performance degradation on bigger files without duplicated headers.

So I think we should: 

1. Get the first line (you already have it as `headers` variable)
2. Compute the new header: `header_map.join(delim)`
3. If both values are different alter the input (current code). Otherwise do nothing.

fortydegrees replied:
Great catch. I've updated this now. It will still fire if transformHeaders is called. This can be fixed by setting a flag i.e. `edited = true` in:

```
if (count > 0) {
	headerName = header + separator + count;
}
```

and then checking for that flag instead of comparing the two arrays. Otherwise it works fine. Let me know if you'd rather this solution.

pokoli commented on papaparse.js:
```diff
@@ -1457,6 +1457,32 @@ License: MIT
+					var header_name = header
+
+					let count = header_count[header] || 0
+					if (count > 0) header_name = header+separator+count;
```
ok, that looks nicer but keep the if statement in separate lines: 

```
if (count > 0) {
    header_name = header+separator+count;
}
```

pokoli commented:
@fortydegrees linting is failing, could you have a look? 

https://github.com/mholt/PapaParse/actions/runs/3463298598/jobs/5783355405

fortydegrees commented:
> @fortydegrees linting is failing, could you have a look?
> 
> https://github.com/mholt/PapaParse/actions/runs/3463298598/jobs/5783355405

Should be fixed now. Thank you for your excellent help in cleaning everything up!

pokoli commented:
@fortydegrees I've merged your work. Thanks you so much!

sethmcleod commented:
Thank you for resolving this @fortydegrees!

@pokoli Would you mind doing another minor version bump so we can use these changes in production?
